### PR TITLE
Cones

### PR DESCRIPTION
**Pull request by valeriaRaffuzzi**
Here's a cone quadratic surface. It's an infinite cone, aligned with the x, y or z axis. 
The input definition is the same as MCNP (need to specify the tangent of the opening angle and the orientation of the cone), compatible with Serpent as well. 

I wrote a unit test and compared my results with Serpent, which agree.

### CODE REVIEW

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+   end select
+
+   ! Load data
+   self % t_sq   = tangent * tangent
+   self % vertex = vertex
+   self % dir = dir
+   call self % setID(id)
+
+   ! Set surface tolerance
+   call self % setTol(TWO * SURF_TOL)
```
Comment (Mikolaj-A-Kowalski) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:188]: It probably should be made consistent with the value in the doc-comment. 

Sadly there is a general problem with the surfaces and surface tolerance I missed when the current geometry was first made. 

I wanted the surface tolerance to be well defined and be approximately a shell of constant 'fuzzy thickness' around the surface, but it sadly will not work with general quadratic surfaces where the 'radius' changes. Hence one cannot get away with the approximation of `2 * r * SURF_TOL` which worked well for cylinders and spheres...   

I think it is fine to leave it as it is but perhaps there should be a comment that will say that 'the thickness of the fuzzy region around the surface will depend on the position on the surface and is not constant'. 

We probably should open an issue about it as well. We need to rethink how the surface tolerance is done.

**Review by Mikolaj-A-Kowalski — COMMENTED** [review]
Thanks for the contribution and sorry that surfaces are bit of the mess in SCONE. 

The only major question I have is if wouldn't it be easier to just include both cones in the surface instead of trying to use only one? I guess it makes it more user friendly but it seems to require considerable complication when evaluating the surface expression or finding the distance.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  !!
+  !! F(r) = (ri - oi)^2 + (rj - oj)^2 - t^2 * (rk - ok)^2 = 0
+  !!
+  !! Where i,j,k are x,y & z axis, and t is the tangent of the cone opening angle
+  !! The cone is aligned along the k axis
+  !!
+  !! The cone can be oriented towards the +k or -k directions; in the input the entry
+  !! orientation should respectively be 1 or -1.
+  !!
+  !! Three diffrent types are avaliable
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:24]:
```suggestion
  !! Three different types are available
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    integer(shortInt), intent(in)           :: id
+    character(*), intent(in)                :: type
+    real(defReal), dimension(3), intent(in) :: vertex
+    real(defReal), intent(in)               :: tangent
+    integer(shortInt), intent(in)           :: dir
+    character(100), parameter :: Here = 'build (cylinder_class.f90)'
+
+    ! Check values
+   if (id < 1) then
+     call fatalError(Here,'Invalid surface id provided. ID must be > 1')
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:152]:
```suggestion
     call fatalError(Here,'Invalid surface id provided. ID must be > 0')
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+  !!
+  !! Test boundary conditions
+  !!
+@Test(cases=[1,2,3])
+  subroutine testBC(this)
+    class(test_cone), intent(inout) :: this
+    real(defReal), dimension(3)     :: r, u, r_pre, u_pre
+
+    ! Set Boundary Contidions
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:171]:
```suggestion
    ! Set Boundary Conditions
```

File: Geometry/geometryStd_class.f90 [hunk 5]
```diff
@@ -414,18 +414,18 @@ function activeMats(self) result(matList)
   end function activeMats
 
   !!
-  !! Descend down the geometry structure untill material is reached
+  !! Descend down the geometry structure until material is reached
   !!
-  !! Requires strting level to be specified.
+  !! Requires starting level to be specified.
   !! It is private procedure common to all movment types in geometry.
```
Comment (ChasingNeutrons) [inline Geometry/geometryStd_class.f90:420]:
```suggestion
  !! It is a private procedure common to all movement types in geometry.
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    u(a) = ZERO
+    u(p1) = -ONE
+    u(p2) = ZERO
+    u = u/norm2(u)
+    ref = TWO
+    @assertEqual(ref, this % surf % distance(r, u), TOL * ref)
+
+    ! Particle going outside
+    @assertEqual(INF, this % surf % distance(r, -u))
+
```
[the remark this answers is not in the document]
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/Tests/cone_test.f90:349]: This should be gone now, thanks to few more checks! 

I also changed the orientation thing, and now one can truncate the cone however he/she wants. Both at the vertex, or at any other height at either side of the vertex.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 1 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90]: The tests were a bit challenges because of this weird tolerance constant through a changing radius. Now everything should work but I agree that it's a bit annoying and suboptimal.

Hopefully this could be discussed in an issue and addressed in another PR!

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+    else
+      d1 = - (b + sqrt(delta)) / a
+      d2 = - (b - sqrt(delta)) / a
+
+    end if
+
+    ! If one the distance found is smaller than the surface tolerance, then the particles
+    ! is on a surface: set that distance to infinity
+    if (d1 < self % surfTol()) d1 = INF
+    if (d2 < self % surfTol()) d2 = INF
```
Comment (Mikolaj-A-Kowalski) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:306]: I am not sure if this approach will work I'm afraid... :-( 
The problem are particles hitting the surface at very shallow angle. They may be well within the tolerance but the distance to the surface may be still quite large. This is why in the e.g. 'cylinder' we need an extra case in the if-else for when a particle is within surface tolerance. 

Sorry... geometry is super annoying.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+    ! Verify if the points found would fall inside the cone
+    testPoint2 = r(self % plane) + d1 * u(self % plane) - self % vertex(self % plane)
+    testRadius = dot_product(testPoint2, testPoint2)
+    maxRadius  = self % hMax * self % hMax * self % tan2
+    if (testRadius > maxRadius) d1 = INF
+
+    testPoint2 = r(self % plane) + d2 * u(self % plane) - self % vertex(self % plane)
+    testRadius = dot_product(testPoint2, testPoint2)
+    maxRadius  = self % hMin * self % hMin * self % tan2
+    if (testRadius > maxRadius) d2 = INF
```
Comment (Mikolaj-A-Kowalski) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:334]: As an optimisation we can switch to the approach similar to the one  used in the `box`... basically it involves calculating the ranges on the line that fall within a particular body (e.g. inside of the cylinder or space between a pair of planes) and then performing the intersections on the ranges. But is is sadly a bit difficult to explain in the short comment ;-( We can talk about it in person perhaps. 

That being said we can leave the explicit approach if it works well and include the alternative as a TODO in the comment.

**Review by Mikolaj-A-Kowalski — COMMENTED** [review]
Looks good and thanks for the changes.
But because it is geometry it makes me super paranoid. Did you run a lot of surface tracking calculations with the cylinders yet? We probably should construct geometries and torture them with the `rayVol` physics package a bit before merging. 
I will try to run some when I have some time.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+    if (id < 1) call fatalError(Here,'Invalid surface id provided. ID must be > 0')
+
+    if (size(vertex) /= 3) then
+      call fatalError(Here,'Vertex needs to have size 3. It has size: '//numToChar(size(vertex)))
+    end if
+
+    if (angle <= ZERO .or. angle >= 90.0_defReal) then
+      call fatalError(Here, 'Opening angle of cone must be in the range 0-90 degrees. &
+                            & It is: '//numToChar(angle))
+    end if
```
[the remark this answers is not in the document]
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:141]: Maybe the error message should be reworded, but the cases of `0` and `90` should be caught as an error.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 7 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:306]: I see, I hadn't thought about that! I will reintroduce the extra if-else case when solving the quadratic, hopefully that will do it!

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 9 above)
[the remark this answers is not in the document]
Comment (Mikolaj-A-Kowalski) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:141]: ... or I should learn to read better :sweat_smile: 
Sorry, I misread the if statement and though these cases are accepted.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 8 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:334]: No that's ok, over the weekend I will have a look at the box and fix this! But I would still need to calculate explicitly whether the particle would intersect the circles? Because in a box or cylinder the two faces are of the same size but this is not true here.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 9 above)
[the remark this answers is not in the document]
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:141]: No worries, I can add an explicit comment saying that the degenerate cases of 0 and 90 are not acceptable.

**Comment by valeriaRaffuzzi** [comment]
> Looks good and thanks for the changes. But because it is geometry it makes me super paranoid 😨 Did you run a lot of surface tracking calculations with the cones yet? We probably should construct geometries and torture them with the `rayVol` physics package a bit before merging. I will try to run some when I have some time.

Thanks a lot for reviewing this, I am a geometry beginner! I ran a couple of cases but not `rayVol` to be fair. I would say hold on with this, I will work on it a bit more during the weekend and fix everything you mentioned, and then do a bit more testing :)

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 8 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90]: Ok, the new distance calculation should be more like box and truncated cylinder now. The only thing I had to add is a check to make sure that the intersection found is on the right half-cone. Otherwise, the segment between two intersections could have been outside of the cone. If an intersection is on the wrong side of the cone, the distance is set to + or - infinity in a way that the intersection 'segment' (or ray) crosses the truncated cone (when here is an actual crossing). This might not be clear, in case let me know!

**Comment by valeriaRaffuzzi** [comment]
> > Looks good and thanks for the changes. But because it is geometry it makes me super paranoid 😨 Did you run a lot of surface tracking calculations with the cones yet? We probably should construct geometries and torture them with the `rayVol` physics package a bit before merging. I will try to run some when I have some time.
> 
> Thanks a lot for reviewing this, I am a geometry beginner! I ran a couple of cases but not `rayVol` to be fair. I would say hold on with this, I will work on it a bit more during the weekend and fix everything you mentioned, and then do a bit more testing :)
> 
> Hopefully at least we're converging to something with these cones!

I think `rayVol` might be broken. I tried to use it on regular JEZ and POPSY and I get an error: `Coordinate list is not placed in the geometry`, in `move`. We should open an issue about that. However, I tested the cones on a few configurations doing ST vs DT and they always agree perfectly! I also tried to make the unit test as varied as I could in terms of distance calculations. Let me know if there is anything else I could do to test this.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 1 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90]: I changed the tolerance to be proportional to the 'mean radius' of the cone now. That is the radius of the cone at mid-height.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+    ! Check values
+    if (id < 1) call fatalError(Here,'Invalid surface id provided. ID must be > 0')
+
+    if (size(vertex) /= 3) then
+      call fatalError(Here,'Vertex needs to have size 3. It has size: '//numToChar(size(vertex)))
+    end if
+
+    if (angle <= ZERO .or. angle >= 90.0_defReal) then
+      call fatalError(Here, 'Opening angle of cone must be in the range 0-90 degrees &
+                            & (extremes excluded). It is: '//numToChar(angle))
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:144]: Maybe my Fortran is getting rusty, shouldn't this be:
```suggestion
                    call fatalError(Here, 'Opening angle of cone must be in the range 0-90 degrees '//&
                            '(extremes excluded). It is: '//numToChar(angle))
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+
+    ! Check values
+    if (id < 1) call fatalError(Here,'Invalid surface id provided. ID must be > 0')
+
+    if (size(vertex) /= 3) then
+      call fatalError(Here,'Vertex needs to have size 3. It has size: '//numToChar(size(vertex)))
+    end if
+
+    if (angle <= ZERO .or. angle >= 90.0_defReal) then
+      call fatalError(Here, 'Opening angle of cone must be in the range 0-90 degrees &
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:143]: Maybe my Fortran is getting rusty, shouldn't this be:
```suggestion
                    call fatalError(Here, 'Opening angle of cone must be in the range 0-90 degrees '//&
                            '(extremes excluded). It is: '//numToChar(angle))
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+                            & (extremes excluded). It is: '//numToChar(angle))
+    end if
+
+    if (hMin >= hMax) call fatalError(Here, 'hMin is greater or equal hMax.')
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:147]:
```suggestion
    if (hMin >= hMax) call fatalError(Here, 'hMin is greater than or equal to hMax.')
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+  pure function boundingBox(self) result(aabb)
+    class(cone), intent(in)     :: self
+    real(defReal), dimension(6) :: aabb
+    real(defReal)               :: maxRadius
+
+    ! Along the axis
+    aabb(self % axis) = self % vertex(self % axis) + self % hMin
+    aabb(self % axis + 3) = self % vertex(self % axis) + self % hMax
+
+    ! On the plane of the basis
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:224]:
```suggestion
    ! On the plane of the base
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+
+  !!
+  !! Evaluate surface expression c = F(r)
+  !!
+  !! See surface_inter for details
+  !!
+  pure function evaluate(self, r) result(c)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal)                           :: c, cMin, cMax, cBasis
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:239]:
```suggestion
    real(defReal)                           :: c, cMin, cMax, cBase
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+    diff = r - self % vertex
+
+    ! Evaluate the expression for the surface of the cone
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tanSquare * diff(self % axis) ** 2
+
+    ! Evaluate the expressions for the two basis of the cone and find the maximum
+    cMin = - diff(self % axis) + self % hMin
+    cMax = diff(self % axis) - self % hMax
+    cBasis = max(cMin, cMax)
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:252]:
```suggestion
    cBase = max(cMin, cMax)
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+
+    ! Find the overall maximum
+    c = max(c, cBasis)
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:255]:
```suggestion
    c = max(c, cBase)
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+  pure function evaluate(self, r) result(c)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal)                           :: c, cMin, cMax, cBasis
+    real(defReal), dimension(3)             :: diff
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Evaluate the expression for the surface of the cone
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:245]: All of the cone is arguably a surface.
```suggestion
    ! Evaluate the expression for the slanted surface of the cone
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 18]
```diff
@@ -0,0 +1,451 @@
...
+
+    ! Calculate quadratic components in the cone
+    a = dot_product(u(self % plane), u(self % plane)) - &
+        self % tanSquare * u(self % axis)**2
+    b = dot_product(diff(self % plane) , u(self % plane)) - &
+        self % tanSquare * diff(self % axis) * u(self % axis)
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tanSquare * diff(self % axis) ** 2
+
+    ! Calculate delta/4
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:296]: What is delta? You mean the discriminant? That also does not appear to be what is calculated here - that would be b^2 / 4 - a*c

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 19]
```diff
@@ -0,0 +1,451 @@
...
+    delta = b * b - a * c
+
+    ! Calculate the distances from the cone surface
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:299]:
```suggestion
    ! Calculate the distances from the slanted surface
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+    ! Calculate the distances from the basis of the cone
+    if (abs(u(self % axis)) > epsilon(ONE)) then    ! Normal intersection
+
+      d1 = (self % hMax - diff(self % axis)) / u(self % axis)
+      d2 = (self % hMin - diff(self % axis)) / u(self % axis)
+
+    else                                            ! Particle parallel to axis: check location
+
+      if (diff(self % axis) > self % hMin .and. diff(self % axis) < self % hMax) then
+        ! Inside the cone: basis intersection segment is -INF:+INF
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:340]: Should this be distance rather than segment?
```suggestion
        ! Inside the cone: base intersection distance is -INF:+INF
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 21]
```diff
@@ -0,0 +1,451 @@
...
+        d1 = -INF
+        d2 = INF
+      else
+        ! Outside the cone: basis intersection segment doesn't exist
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:344]:
```suggestion
        ! Outside the cone: base intersection distance doesn't exist
```

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 [hunk 22]
```diff
@@ -0,0 +1,451 @@
...
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal), dimension(3), intent(in) :: u
+    logical(defBool)                        :: halfspace
+    real(defReal), dimension(3)             :: diff, norm
+    real(defReal)                           :: proj
+
+    diff = r - self % vertex
+
+    ! Check the location of the particle, i.e., basis or cone surface, to calculate
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/QuadSurfaces/cone_class.f90:402]:
```suggestion
    ! Check the location of the particle, i.e., base or cone surface, to calculate
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 23]
```diff
@@ -0,0 +1,481 @@
...
+    u     = ZERO
+    u(a)  = ONE
+    u(p1) = -ONE
+    u = u / norm2(u)
+
+    ! At the surface
+    @assertFalse(this % surf % halfspace(r, u))
+
+    ! Out within tolerance
+    ! NOTE THAT the tolerance is very sensitive: if using eps = HALF * TOL, the
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:235]:
```suggestion
    ! NOTE that the tolerance is very sensitive: if using eps = HALF * TOL, the
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 24]
```diff
@@ -0,0 +1,481 @@
...
+
+    ! Out outside tolerance
+    eps = -TWO * tolerance
+    @assertTrue(this % surf % halfspace(r + eps*u, u))
+
+    ! Well Outside
+    eps = -HALF
+    @assertTrue(this % surf % halfspace(r + eps*u, u))
+
+    ! Well withn
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:248]:
```suggestion
    ! Well within
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 25]
```diff
@@ -0,0 +1,481 @@
...
+    eps = HALF
+    @assertFalse(this % surf % halfspace(r + eps*u, u))
+
+    ! Tangent particle should be outside
+    u     = ZERO
+    u(p2) = -ONE
+    u = u /norm2(u)
+    @assertTrue(this % surf % halfspace(r, u))
+
+    ! Choose point on one of the basis of the cone, moving out
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:258]:
```suggestion
    ! Choose point on one of the bases of the cone, moving out
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 26]
```diff
@@ -0,0 +1,481 @@
...
+
+    ! Get surface tolerance
+    tolerance = this % surf % surfTol()
+
+    ! Set axis and plane axis indices
+    a = this % axis
+    p1 = this % plane(1)
+    p2 = this % plane(2)
+
+    ! **Outside the cone
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:311]:
```suggestion
    ! Outside the cone
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 27]
```diff
@@ -0,0 +1,481 @@
...
+    r(p2) = ONE
+
+    ! Impact at an angle, direction of flight on an axis
+    u(a)  = ONE
+    u(p1) = ZERO
+    u(p2) = ZERO
+    ref   = 4.0_defReal
+    @assertEqual(ref, this % surf % distance(r, u), TOL * ref)
+
+    ! Impacting the basis
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:323]:
```suggestion
    ! Impacting the bases
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 28]
```diff
@@ -0,0 +1,481 @@
...
+    @assertEqual(INF, this % surf % distance(r, u))
+
+    ! Going in the opposite direction
+    u(a)  = -ONE
+    u(p1) = ZERO
+    u(p2) = ZERO
+    u   = u/norm2(u)
+    @assertEqual(INF, this % surf % distance(r, u))
+
+    ! Parallel to the basis
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:345]:
```suggestion
    ! Parallel to the bases
```

File: Geometry/Surfaces/Tests/cone_test.f90 [hunk 29]
```diff
@@ -0,0 +1,481 @@
...
+
+    ! Parallel to axis
+    u(a)  = ONE
+    u(p1) = ZERO
+    u(p2) = ZERO
+    u = u/norm2(u)
+    ref = 3.0_defReal
+    @assertEqual(ref, this % surf % distance(r, u), TOL * ref)
+
+    ! Parallel to basis
```
Comment (ChasingNeutrons) [inline Geometry/Surfaces/Tests/cone_test.f90:404]:
```suggestion
    ! Parallel to bases
```

File: Geometry/geometryStd_class.f90 [hunk 30]
```diff
@@ -419,18 +419,18 @@ function activeMats(self) result(matList)
   end function activeMats
 
   !!
-  !! Descend down the geometry structure untill material is reached
+  !! Descend down the geometry structure until material is reached
   !!
-  !! Requires strting level to be specified.
-  !! It is private procedure common to all movment types in geometry.
+  !! Requires starting level to be specified.
+  !! It is private procedure common to all movement types in geometry.
```
Comment (ChasingNeutrons) [inline Geometry/geometryStd_class.f90:425]:
```suggestion
  !! It is a private procedure common to all movement types in geometry.
```

File: docs/User Manual.rst [hunk 31]
```diff
@@ -517,6 +517,23 @@ Example: ::
 
       billy { id 92; type xCylinder; origin (0.0 0.0 9.0); radius 4.8; }
 
+* cone: cone aligned with x, y or z axis, and truncated arbitrarily on both sides. 
+  The input type has to be ``xCone``, ``yCone`` or ``zCone``. The orientation of the
```
Comment (ChasingNeutrons) [inline docs/User Manual.rst:521]: Orientation is ambiguous with x, y, and z
```suggestion
  The input type has to be ``xCone``, ``yCone`` or ``zCone``. The gradient of the
```

File: docs/User Manual.rst [hunk 32]
```diff
@@ -517,6 +517,23 @@ Example: ::
...
+    surface. [degrees]
+  - hMin: coordinate along the cone axis corresponding to the position of the
+    cone lower basis with respect to the vertex. It could be positive or negative. [cm]
+  - hMax: coordinate along the cone axis corresponding to the position of the
+    cone upper basis with respect to the vertex. It could be positive or negative.
+    Must be of the same sign as hMin, and always larger. [cm]
+
+Example: ::
+
+      billy { id 92; type xCone; vertex (1.1 4.0 2.98); angle 30; hMin 5.0; hMax 15.0; }
```
Comment (ChasingNeutrons) [inline docs/User Manual.rst:535]:
```suggestion
      connor { id 92; type xCone; vertex (1.1 4.0 2.98); angle 30; hMin 5.0; hMax 15.0; }
```

**Review by ChasingNeutrons — APPROVED** [review]
All small things I think. Should be straightforward.

File: docs/User Manual.rst (hunk 32 above)
Comment (valeriaRaffuzzi) [reply docs/User Manual.rst:535]: Did you change name to Connor because it sounds like cone?

Comment (ChasingNeutrons) [reply docs/User Manual.rst:535]: Yes

Comment (valeriaRaffuzzi) [reply docs/User Manual.rst:535]: Fair enough. I love bad humor

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 18 above)
Comment (ChasingNeutrons) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:296]: Sure, but it isn't clear what delta is. If you say 'calculate delta/4' there is no way for someone reading afterwards to understand what you are referring to.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 20 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:340]: I think segment is ok: each couple of distances (base1 and base2, or slanted1 and slanted2) identifies 2 points that create a segment. The intersection between the two segments is what is actually internal to the cone. This is what the code is doing, and I think the docs are consistent with the cylinder and box where this approach is used.

Comment (ChasingNeutrons) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90]: I would then say the segment lies between -INF:INF. It is not clear what a segment being -INF:INF means.

File: Geometry/Surfaces/QuadSurfaces/cone_class.f90 (hunk 18 above)
Comment (valeriaRaffuzzi) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90:296]: Should I call it discriminant instead? Is this what you mean?

Comment (ChasingNeutrons) [reply Geometry/Surfaces/QuadSurfaces/cone_class.f90]: Yes, that would be easier to follow.
